handlers/validator.go

```diff
@@ -1541,21 +1547,116 @@ func ValidatorHistory(w http.ResponseWriter, r *http.Request) {
+					OR participated_sync > 0 
+					OR proposed_blocks > 0 
+					OR missed_blocks > 0 
+					OR orphaned_blocks > 0 
+			);`, index)
+		if err != nil {
+			logger.Errorf("error retrieving lastActionDay for validator-history: %v", err)
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+			return
+		}
+		lastActionEpoch := (lastActionDay + 1) * utils.EpochsPerDay()
+		// if the validator had some duties after the exit epoch we calculate how many extra epochs we have to check after the exit epoch
+		if lastActionEpoch > currentEpoch {
+			extraEpochs = protomath.MinU64(lastActionEpoch, services.LatestEpoch()-1) - currentEpoch
+		}
 	}
 
-	var validatorHistory []*types.ValidatorHistory
+	tableData := make([][]interface{}, 0)
 
-	g := new(errgroup.Group)
+	if extraEpochs > 0 {
+		startEpoch := currentEpoch + 1
+		endEpoch := startEpoch + extraEpochs
+		withdrawalMap, incomeDetails, err := getWithdrawalAndIncome(index, startEpoch, endEpoch)
+		if err != nil {
+			return
+		}
+
+		// if there are additional epochs with duties we have to go through all of them as there can be gaps (after the exit before the duty)
+		for i := endEpoch; i >= startEpoch && len(tableData) < pageLength; i-- {
+			if incomeDetails[index] == nil || incomeDetails[index][i] == nil {
+				continue
+			}
+			// for paging we skip the first X epochs with duties
+			if start > 0 {
+				start--
+				continue
+			}
+			tableData = append(tableData, icomeToTableData(i, incomeDetails[index][i], withdrawalMap[i], currency))
+		}
+
+	}
+
+	extraItemLength := len(tableData)
+	// if we have already found enough items we can skip the 'normal' step.
+	if extraItemLength < pageLength {
+		startEpoch := currentEpoch - start - uint64(pageLength-1-extraItemLength) // we only get the exact number of epochs to get to the page length
+		endEpoch := currentEpoch - start
+		if startEpoch > endEpoch { // handle underflows of startEpoch
+			startEpoch = 0
+		}
+
+		withdrawalMap, incomeDetails, err := getWithdrawalAndIncome(index, startEpoch, endEpoch)
+		if err != nil {
+			return
+		}
+
+		for i := endEpoch; i >= startEpoch && len(tableData) < pageLength; i-- {
+			if incomeDetails[index] == nil || incomeDetails[index][i] == nil {
+				if i <= endEpoch {
+					tableData = append(tableData, []interface{}{
+						utils.FormatEpoch(i),
+						"pending...",
+						template.HTML(""),
+						template.HTML(""),
+					})
+				}
+				continue
+			}
+			tableData = append(tableData, icomeToTableData(i, incomeDetails[index][i], withdrawalMap[i], currency))
+		}
+	}
+
+	if len(tableData) == 0 {
+		tableData = append(tableData, []interface{}{
+			template.HTML("Validator no longer active"),
+		})
+	}
 
-	startEpoch := currentEpoch - start - 9
-	endEpoch := currentEpoch - start
-	if startEpoch > endEpoch { // handle underflows of startEpoch
-		startEpoch = 0
+	data := &types.DataTableResponse{
+		Draw:            draw,
+		RecordsTotal:    totalCount,
+		RecordsFiltered: totalCount,
```

Comment:
The calculation for `totalCount` now is incorrect, is it not?
It is based on the validator's exit epoch but now we know that there might be entries in the history even after that.

I suggest to base the calculation on `lastActionEpoch` or the current epoch (whichever is smaller).

Thoughts?

Reply:
right totalCount could now be wrong (if < 100 entries) ... but as we could have gaps I needed to take a bit more complex approach